cmd/vicadmin/server.go

```diff
@@ -499,12 +507,41 @@ func (s *server) tailFiles(res http.ResponseWriter, req *http.Request, names []s
 	}
 }
 
+// deriveDisplayError takes a vSphere session error and returns
+// an error string that is safe to display on the vicadmin page.
+func deriveDisplayError(err error) string {
+	var displayErrStr string
+
+	if err != nil {
+		switch err := err.(type) {
+		case session.SDKURLError:
+			displayErrStr = fmt.Sprintf("SDK URL could not be parsed: %s", err.Err)
+		case session.SoapClientError:
+			displayErrStr = "unable to obtain a vim client"
+		case session.UserPassLoginError:
+			displayErrStr = "unable to log in with username and password"
+		default:
+			displayErrStr = err.Error()
+		}
+	}
+
+	return displayErrStr
+}
+
 func (s *server) index(res http.ResponseWriter, req *http.Request) {
 	defer trace.End(trace.Begin(""))
 	ctx := context.Background()
 	sess, err := s.getSessionFromRequest(req)
 	v := vicadmin.NewValidator(ctx, &vchConfig, sess)
 
+	if sess == nil {
+		// We're unable to connect to vSphere, so display an error message
+		displayErrStr := deriveDisplayError(err)
+		v.VCHIssues = template.HTML(fmt.Sprintf(
+			"<span class=\"error-message\">%s Error: %s</span>\n",
+			v.Hostname+VCHUnreachableMsg, displayErrStr))
```

Comment:
I feel this improves readability by separating the html arg from the other args.